cpp/examples/tpch/q1.cpp

```diff
@@ -0,0 +1,148 @@
+    return charge;
+}
+
+int main(int argc, char const** argv) {
+    auto args = parse_args(argc, argv);
+
+    // Use a memory pool
+    auto resource = create_memory_resource(args.use_memory_pool);
+    rmm::mr::set_current_device_resource(resource.get());
+
+    Timer timer;
+
+    // 1. Read out the `lineitem` table from parquet file
+    auto shipdate_ref = cudf::ast::column_reference(5);
+    auto shipdate_upper = cudf::timestamp_scalar<cudf::timestamp_D>(days_since_epoch(1998, 9, 2), true);
+    auto shipdate_upper_literal = cudf::ast::literal(shipdate_upper);
+    auto shipdate_pred = std::make_unique<cudf::ast::operation>(
+        cudf::ast::ast_operator::LESS_EQUAL,
+        shipdate_ref,
+        shipdate_upper_literal
+    );
+    auto lineitem = read_parquet(
+        args.dataset_dir + "lineitem/part-0.parquet", 
+        {"l_returnflag", "l_linestatus", "l_quantity", "l_extendedprice", "l_discount", "l_shipdate", "l_orderkey", "l_tax"},
+        std::move(shipdate_pred)
+    );
+
+    // 2. Calculate the discount price and charge columns and append to lineitem table
+    auto disc_price = calc_disc_price(lineitem);
+    auto charge = calc_charge(lineitem, disc_price);
+    auto appended_table = lineitem->append(disc_price, "disc_price")->append(charge, "charge");
+
+    // 3. Perform the group by operation
+    auto groupedby_table = apply_groupby(
+        appended_table, 
+        groupby_context_t {
+            {"l_returnflag", "l_linestatus"},
+            {
+                {
+                    "l_extendedprice", 
+                    {
+                        {cudf::aggregation::Kind::SUM, "sum_base_price"}, 
+                        {cudf::aggregation::Kind::MEAN, "avg_price"}    
+                    }
+                },
+                {
+                    "l_quantity", 
+                    {
+                        {cudf::aggregation::Kind::SUM, "sum_qty"},
+                        {cudf::aggregation::Kind::MEAN, "avg_qty"}
+                    }
+                },
+                {
+                    "l_discount", 
+                    {
+                        {cudf::aggregation::Kind::MEAN, "avg_disc"},
+                    }
+                },
+                {
+                    "disc_price", 
+                    {
+                        {cudf::aggregation::Kind::SUM, "sum_disc_price"},
+                    }
+                },
+                {
+                    "charge", 
+                    {
+                        {cudf::aggregation::Kind::SUM, "sum_charge"}, 
+                        {cudf::aggregation::Kind::COUNT_ALL, "count_order"}
+                    }
+                },
+            }
+        }
+    );
+
+    // 4. Perform the order by operation
+    auto orderedby_table = apply_orderby(groupedby_table, {"l_returnflag", "l_linestatus"}, {cudf::order::ASCENDING, cudf::order::ASCENDING});
+
+    timer.print_elapsed_millis();
```

Comment:
Do we need to ensure that the device (or stream) is synchronized before ending the timer? There are probably implicit synchronizations already occurring, but it might be good to check. We would like libcudf to be fully stream-ordered (no implicit syncs except where required for correctness) over time.

Reply:
@bdice , sounds good, can you please point me to how to do this ? Like using `cudaStreamSynchronize` ?

Reply:
You can do this by explicitly passing the stream you are using to `cudf::sort_by_key` inside this function and then do a `stream.synchronize()`